FileIO/MPI_IO/NodePartitionedMeshReader.cpp

```diff
@@ -0,0 +1,521 @@
+
+        if(part_id == 0)
+            setNodes(nodes, mesh_nodes, glb_node_ids);
+        else
+            MPI_Send(nodes.data(), static_cast<int>(_mesh_info.nodes),
+                _mpi_node_type, part_id, message_tag, _mpi_comm);
+    }
+    else if(part_id > 0 && _mpi_rank == part_id)
+    {
+        MPI_Recv(nodes.data(), static_cast<int>(_mesh_info.nodes),
+            _mpi_node_type, 0, message_tag, _mpi_comm, MPI_STATUS_IGNORE);
+        setNodes(nodes, mesh_nodes, glb_node_ids);
+    }
+
+    return true;
+}
+
+bool NodePartitionedMeshReader::readCastElemsASCII(std::ifstream& is_elem,
+    const int part_id, const std::size_t data_size, const bool process_ghost,
+    const std::vector<MeshLib::Node*> &mesh_nodes,
+    std::vector<MeshLib::Element*> &mesh_elems) const
+{
+    int const message_tag = 0;
+
+    // MPI_Send/Recv can handle only int. Check for overflow.
+    if (!is_safely_convertable<std::size_t, int>(data_size))
+    {
+        ERR("Too large number of elements to read.");
+        return false;
+    }
+
+    std::vector<unsigned long> elem_data(data_size);
+    if(_mpi_rank == 0)
+    {
+        readElementASCII(is_elem, elem_data, process_ghost);
+
+        if(part_id == 0)
+        {
+            if(!process_ghost)
+                mesh_elems.resize(
+                    _mesh_info.regular_elements + _mesh_info.ghost_elements);
+            setElements(mesh_nodes, elem_data, mesh_elems, process_ghost);
+        }
+        else
+            MPI_Send(elem_data.data(), static_cast<int>(data_size), MPI_LONG,
+                part_id, message_tag, _mpi_comm);
+    }
+    else if(part_id > 0 && _mpi_rank == part_id)
+    {
+        MPI_Recv(elem_data.data(), static_cast<int>(data_size), MPI_LONG,
+            0, message_tag, _mpi_comm, MPI_STATUS_IGNORE);
+
+        if(!process_ghost)
+            mesh_elems.resize(
+                _mesh_info.regular_elements + _mesh_info.ghost_elements);
+        setElements(mesh_nodes, elem_data, mesh_elems, process_ghost);
+    }
+
+    return true;
+}
+
+MeshLib::NodePartitionedMesh* NodePartitionedMeshReader::readASCII(
+    const std::string &file_name_base)
+{
+    std::ifstream is_cfg;
+    std::ifstream is_node;
+    std::ifstream is_elem;
+
+    bool file_opened = false;
+    if(_mpi_rank == 0)
+    {
+        file_opened = openASCIIFiles(file_name_base, is_cfg, is_node, is_elem);
+
+        if(!file_opened)
+        {
+            is_cfg.close();
+            is_node.close();
+            is_elem.close();
+        }
+    }
+
+    MPI_Bcast(&file_opened, 1, MPI_INT, 0, _mpi_comm);
+
+    if(!file_opened)
+        return nullptr;
+
+    MeshLib::NodePartitionedMesh* np_mesh = nullptr;
+    std::vector<MeshLib::Node*> mesh_nodes;
+    std::vector<unsigned long> glb_node_ids;
+    std::vector<MeshLib::Element*> mesh_elems;
+
+    for(int i = 0; i < _mpi_comm_size; i++)
+    {
+        if(_mpi_rank == 0)
+        {
+            INFO("-->Parallel reading the partitioned mesh: ");
```

Comment:
Yes, it is. The ASCII input is for test purpose. With that only rank 0 takes charge of reading data and cast them. I wanted to remove this functionality before because it is less usage but it is gained more coding and review efforts.  On the contrary binary input is purely parallelized for practical application.